synapse/api/filtering.py

```diff
@@ -202,7 +201,7 @@ def check_valid_filter(self, user_filter_json: JsonDict) -> None:
 
 
 # Filters work across events, presence EDUs, and account data.
-FilterEvent = TypeVar("FilterEvent", EventBase, UserPresenceState, JsonDict)
+FilterEvent = Union["FilterEvent", EventBase, UserPresenceState, JsonDict]
```

Comment:
I think I used a `TypeVar` here because the APIs spit out the same thing they get in, so by switching to a `Union` we're widening the potential return values.

Reply:
This seems to have blown up in my face. Odd, could have sworn it was fine locally.

Reply:
It might make sense to do everything except the filter file for now and just get that done? 🤷

Reply:
As written on develop, I think a function that expects `x: List[FilterEvent]` expects x to be one of  `List[EventBase]`, `List[UserPresenceState]`, or `List[JsonDict]`, expects to be given a homogeneous list.

However this test passes in a heterogeneous list:

https://github.com/matrix-org/synapse/blob/e37218a4e4a5710196912469075e42564210aa85/tests/api/test_filtering.py#L535-L572

Maybe the test can be rewritten to avoid this.... though if I can't see anything obvious, I'll take your advice and exclude this file.

Reply:
> As written on develop, I think a function that expects `x: List[FilterEvent]` expects x to be one of `List[EventBase]`, `List[UserPresenceState]`, or `List[JsonDict]`, expects to be given a homogeneous list.
> 
> However this test passes in a heterogeneous list of `events` to `_check_event_relations`:

Correct; I think the tests and what the code actually uses it for don't quite line up. (Or at least no longer do.) I think it only ever has to accept a homogeneous list, as you said.

Reply:
c9a411611b is an attempt at this. There are a few commits though, it might be worth going through them sequentially.